homeassistant/components/camera/amcrest.py

```diff
@@ -64,26 +82,84 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
             notification_id=NOTIFICATION_ID)
         return False
 
-    add_devices([AmcrestCam(config, data)])
+    add_devices([AmcrestCam(hass, config, data)])
     return True
 
 
 class AmcrestCam(Camera):
     """An implementation of an Amcrest IP camera."""
 
-    def __init__(self, device_info, data):
+    def __init__(self, hass, device_info, data):
         """Initialize an Amcrest camera."""
         super(AmcrestCam, self).__init__()
+        self._base_url = '%s://%s:%s/cgi-bin' % (
+            'http',
+            device_info.get(CONF_HOST),
+            device_info.get(CONF_PORT)
+        )
         self._data = data
+        self._hass = hass
         self._name = device_info.get(CONF_NAME)
         self._resolution = RESOLUTION_LIST[device_info.get(CONF_RESOLUTION)]
+        self._stream_source = STREAM_SOURCE_LIST[
+            device_info.get(CONF_STREAM_SOURCE)
+        ]
+        self._token = self._auth = aiohttp.BasicAuth(
+            device_info.get(CONF_USERNAME),
+            password=device_info.get(CONF_PASSWORD)
+        )
+        self._websession = async_create_clientsession(hass)
 
     def camera_image(self):
         """Return a still image reponse from the camera."""
         # Send the request to snap a picture and return raw jpg data
         response = self._data.camera.snapshot(channel=self._resolution)
         return response.data
 
+    @asyncio.coroutine
+    def handle_async_mjpeg_stream(self, request):
+        """Return an MJPEG stream."""
+        # The snapshot implementation is handled by the parent class
+        if self._stream_source == STREAM_SOURCE_LIST['snapshot']:
+            yield from super().handle_async_mjpeg_stream(request)
+            return
+
+        # Otherwise, stream an MJPEG image stream directly from the camera
+        streaming_url = '%s/mjpg/video.cgi?channel=0&subtype=%d' % (
+            self._base_url,
+            self._resolution
+        )
+
+        stream = None
+        response = None
+        try:
+            with async_timeout.timeout(TIMEOUT, loop=self.hass.loop):
+                stream = yield from self._websession.get(
+                    streaming_url,
+                    auth=self._token,
+                    timeout=TIMEOUT
+                )
+            response = web.StreamResponse()
+            response.content_type = stream.headers.get(CONTENT_TYPE_HEADER)
+
+            yield from response.prepare(request)
+
+            while True:
+                data = yield from stream.content.read(16384)
+                if not data:
+                    break
+                response.write(data)
+
+        except (asyncio.TimeoutError, aiohttp.errors.ClientError):
+            _LOGGER.exception("Error on %s", streaming_url)
+            raise HTTPGatewayTimeout()
+
+        finally:
+            if stream is not None:
+                self.hass.async_add_job(stream.release())
```

Comment:
Change this to `yield from stream.close()` or you'll run out of memory.

(in another PR we should probably make a generic mjpeg proxy that only requires a ClientResponse object as this code can be shared among camera implementations)